### PR TITLE
Language aware mapping

### PR DESCRIPTION
**Pull request by chreman**
Please only review after #271 could be merged, the diffs will be less.
This PR introduces language-awareness at various stages of the pipeline:
* language is detected in the preprocessing stage and added to the metadata
* the corpus utilized for creating the document representation is now created language-aware: each document is now cleaned from stopwords of each language; each document is now stemmed according to its language
* building on an improved feature representation, language can now be added to the distance matrix before clustering, which will lead to clusters that are additionally separated by language
* it also makes and exception for German when lowercasing the bubbletitles

Additionally, language-specific logging has been added at various stages.

### CODE REVIEW

**Comment by pkraker** [comment]
Thanks. I did run a snapshot test (base-snapshot.R) just now on Ubuntu 16.04 and it turned up a sizeable amount of differences between the actual and the expected result (see below).

```
Table: Non-identical attributes

|                            |
|:---------------------------|
|No non-identical attributes |

            var.x          var.y   n NAs
1        relation       relation   0   0
2      identifier     identifier   0   0
3           title          title   0   0
4  paper_abstract paper_abstract   0   0
5    published_in   published_in   0   0
6            year           year   0   0
7    subject_orig   subject_orig   0   0
8         subject        subject   0   0
9         authors        authors   0   0
10           link           link   0   0
11       oa_state       oa_state   0   0
12            url            url   0   0
13      relevance      relevance   0   0
14              x              x 100   0
15              y              y 100   0
16       area_uri       area_uri  31   0
17 cluster_labels cluster_labels  41   0
18           area           area  41   0
```

**Comment by pkraker** [comment]
The snapshot test also produced a number of warnings that I have not seen before:
```
Warning: Variables excluded for having levels > max_levels: id
Warning in as.dist.default(x) : non-square matrix
```

File: server/preprocessing/other-scripts/summarize.R [hunk 1]
```diff
@@ -31,27 +31,44 @@ filter_out <- function(ngrams, stops){
   return (tokens)
 }
 
-create_cluster_labels <- function(clusters, metadata_full_subjects, weightingspec, top_n, stops, taxonomy_separator="/") {
+create_cluster_labels <- function(clusters, metadata, weightingspec, lang,
+                                  top_n, stops, taxonomy_separator="/") {
+  nn_corpus <- get_cluster_corpus(clusters, metadata, stops, taxonomy_separator)
+  tolower_flag <- if (lang == "german") FALSE else TRUE
```
[the remark this answers is not in the document]
Comment (pkraker) [reply server/preprocessing/other-scripts/summarize.R]: If there are different uppercase versions for the same lowercase word, I would use the one that has more occurrences. If they have the same number of occurrences, I would use the one that has more uppercase letters. If both conditions are the same, I would choose a random word :)

**Comment by chreman** [comment]
* The differences in the snapshot tests are expected, as they are based on snapshots from master, I should've explained this better. The idea is that when workflow changes are introduced, their effects can be traced a bit better. Once new features are merged to master, new snapshots will be created based on the new workflow.
* The warnings in the snapshots are OK, as `id` is unique it's unique count is 100 and above a certain threshold in the compare-function.
* I also added a few replies to your line-comments

File: server/preprocessing/resources/german.stop [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+da
+damit
+dann
+der
+den
+des
+dem
+die
+das
+daß
```
[the remark this answers is not in the document]
Comment (chreman) [reply server/preprocessing/resources/german.stop]: That may indeed be the case, I will look for a more recent stopwords-dict.

File: server/preprocessing/other-scripts/features.R [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+create_corpus <- function(metadata, text, stops) {
+  valid <- getStemLanguages()
+  text["language"] <- unlist(lapply(metadata$lang_detected,
+                    function(x) {
+                      if (x %in% valid) {
+                        x
+                      } else {
+                        "english"
+                      }}))
+  mapping <- list(content = "content", id = "id", language = "language")
```
[the remark this answers is not in the document]
Comment (chreman) [reply server/preprocessing/other-scripts/features.R]: Yes, this applies to stopwords-removal and stemming for now.

File: server/preprocessing/other-scripts/vis_layout.R [hunk 4]
```diff
@@ -3,68 +3,84 @@ library(GMD)
...
+library(textcat)
 library(proxy)
 library(SnowballC)
 library(jsonlite)
 library(parfossil)
 library(doParallel)
 library(stringi)
 library(stringdist)
+library(plyr)
+library(onehot)
```
[the remark this answers is not in the document]
Comment (chreman) [reply server/preprocessing/other-scripts/vis_layout.R]: Done.

File: server/preprocessing/other-scripts/summarize.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (chreman) [reply server/preprocessing/other-scripts/summarize.R]: This functionality has been partially adressed also by the function in L78 in `preprocess.R`, `candidates = mapply(conditional_lowercase, candidates, metadata$lang_detected[missing_subjects])` and the according function `conditional_lowercase`. This is hardcoded for now and neither elegant nor scalable, but it works for German.

[the remark this answers is not in the document]
Comment (pkraker) [reply server/preprocessing/other-scripts/summarize.R]: Agreed, see #284.

File: server/preprocessing/resources/german.stop (hunk 2 above)
[the remark this answers is not in the document]
Comment (pkraker) [reply server/preprocessing/resources/german.stop]: Thanks!

**Comment by chreman** [comment]
1) Fixed the erroneous debug statement
2) This warning may arise from the nested tryCatch which attempts to find a specific additional stopwords file, in this case because it is run from a test.R, the error arises because it has to go two folder hierarchies up. This implementation is definitely not perfect, because it covers a) the difference between test & production (resulting in two different folder levels) and b) the case of existence/non-existence of a stopwords file
3) Yes, discrepancies may arise because the language reported in that statement is what is being detected by textcat. This may be because for example in base an entry has `dclang:ger`, but still may contain a bilingual abstract eng/ger. We may need to evaluate how many bilingual edge-cases exist, but I still believe that it improves our bottomline.

**Comment by pkraker** [comment]
Awesome, this is now ready for merging. I have recorded the Windows-only issue with this solution here: #287

**Pull request merged**